pinot-core/src/main/java/org/apache/pinot/core/data/recordtransformer/NullValueTransformer.java

```diff
@@ -33,17 +36,35 @@ public NullValueTransformer(Schema schema) {
 
   @Override
   public GenericRow transform(GenericRow record) {
+    Set<String> nullColumnNamesSet = null;
+
+    // Clear out the 'null_fields' value in case the Generic row is reused
+    record.putField(NULL_FIELDS, null);
+
     for (FieldSpec fieldSpec : _fieldSpecs) {
       String fieldName = fieldSpec.getName();
       // Do not allow default value for time column
       if (record.getValue(fieldName) == null && fieldSpec.getFieldType() != FieldSpec.FieldType.TIME) {
+        if (nullColumnNamesSet == null) {
+          nullColumnNamesSet = new HashSet<>();
+        }
+
+        // Only handle null columns for non-virtual columns
+        if (!fieldSpec.isVirtualColumn()) {
+          nullColumnNamesSet.add(fieldName);
+        }
+
         if (fieldSpec.isSingleValueField()) {
           record.putField(fieldName, fieldSpec.getDefaultNullValue());
         } else {
           record.putField(fieldName, new Object[]{fieldSpec.getDefaultNullValue()});
         }
       }
     }
+
+    if (nullColumnNamesSet != null) {
+      record.putField(NULL_FIELDS, String.join(",", nullColumnNamesSet));
```

Comment:
I think that we can directly store `Set` since `GenericRow` gets `Object` as value instead of keep repeating `join & parse`?

Reply:
I think the concern is the size of a Java object which is potentially bigger in this case. 

One optimization here is that if we have a fixed order of column names (in the FieldSpec), we can simply use a BitSet to keep track of null columns (which will typically be around 64 bytes ish). This can save a lot of effort in other parts of the code base as well.